accepted/0015-virtual-camera-support.md

```diff
@@ -0,0 +1,69 @@
+* [Wirecast includes virtual camera support](http://www.telestream.net/pdfs/user-guides/Wirecast-8-User-Guide-Windows.pdf) on both Windows and Mac.
+* [Webcamoid](https://webcamoid.github.io/) is an existing open-source cross-platform virtual camera application, which may have some useful tidbits when investigating implementation details for each platform.
+
+# Detailed design
+
+Feature Set:
+
+* A button is added to the Output Controls dock labeled "Start Virtual Camera", which will toggle the start/stop status of the virtual camera output. While virtual camera output is active, the label should read 'Stop Virtual Camera".
+* The virtual camera appears in the system listed as "OBS Virtual Camera".
+* The virtual camera outputs at the same resolution and framerate as the OBS rendered output.
+* A section is added to OBS settings providing the following settings:
+    * A checkbox to indicate if the user wishes to start the virtual camera output automatically when OBS starts
+    * A checkbox to flip the virtual output horizontally
+* In the Output settings tab, the warning that appears when outputs are active is augmented to list all active outputs so that users know which outputs they need to disable in order to edit output settings.
+* When the virtual camera output is not active, the virtual camera device shows a static image to consumers indicating to the consumer that output is not started. This image meets four requirements:
+    * It allows the consumer to smoothly switch from non-active to active content without having to worry about what order the applications are started in.
+    * It informs the user that the output is not started yet, and needs to be started before output will be sent from OBS.
+    * It is aesthetically pleasing such that, if the output is accidentally sent to viewers (for example, if the user fails to start the virtual output before joining a Zoom call), it is not unnecessarily gaudy or technical.
+    * It requires minimal localization, or provides a way to localize any displayed text on the image
+* The auto-config tool is updated to include "outputting to a virtual camera" as their primary use case, with reasonable recommendations. Since this doesn't require testing encoding or bandwidth settings, the resolution should be set to 1920x1080 and frame rate set to 30FPS.
+* A new command line flag `--startvirtualcam` is added, which starts the virtual cam automatically when OBS is started.
+
+## Platform specific implementations
+
+* On **Windows**, the implementation of this plugin should leverage [libdshowcapture](https://github.com/obsproject/libdshowcapture). OBS already depends on libdshowcapture, so using this existing library service limits the need to add additional dependecies.
+* On **macOS**, OBS will need a plugin to output over IPC to a CoreMediaIO DAL plugin that is registered on the system upon OBS install. This will require repackaging the installer as a `.pkg` instead of a compressed `.app`.
```

Comment:
As an alternative to this, when the user first starts the virtual camera, we could potentially prompt to install the DAL plugin at that time. We'd need to show a password prompt to be able to have access to write to `/Library`.

Reply:
How is that better than installing it at OBS install time? It's already elevated at that point and cuts out an extra step.

Reply:
On macOS, since OBS distributes a `.app` bundle inside a `.dmg` image, we don't have elevated privileges at install time currently. When the user first opens OBS, macOS prompts for permission for things that require elevated approval/privileges (e.g. Microphone, Camera, etc). A common pattern I've seen recently for apps that need additional files installed is that they distribute as a `.app` but have menu actions to install things with elevated privileges. This allows the end user to decide to what extent they want to give an app access to their system.

Example from JetBrains' IDEs
![image](https://user-images.githubusercontent.com/218876/81997855-cefe5a00-9605-11ea-9eb7-2d6835e89436.png)

Example from VSCode
![image](https://user-images.githubusercontent.com/218876/81997934-0836ca00-9606-11ea-83ff-b70c9251a5fb.png)

For me, I'm generally suspicious when apps require a `.pkg` installer since it's harder to know exactly what they're installing on my system. I try not to install apps via `.pkg` if I can help it. `.pkg` installers can install sneaky things that break security like [Zoom did](https://techcrunch.com/2019/07/10/apple-silent-update-zoom-app/). Maybe others feel similar to me, or maybe they don't care 🤷

But it probably doesn't really matter either way. I bet OBS has enough notoriety that it won't really affect how people perceive it. And if we believe that the virtual camera is going to be used by the majority of OBS users, then a `.pkg` is definitely the right way to go.

Reply:
> A common pattern I've seen recently for apps that need additional files installed is that they distribute as a .app but have menu actions to install things with elevated privileges.

IMO this is absolutely miserable UX. I'm fairly confident that the wide majority of people will absolutely not care if it's a `.pkg` or an `.app`, but they _will_ care if it works out of the box or not.

The Mac build of OBS was distributed as a `.pkg` for years with no objections, and it was only switched to a `.app` with version 24.0.0 because it didn't _need_ to be a `.pkg`. Shipping a `.pkg` would just be going back to the way things used to be, except now it's properly justified.

Finally, unlike Zoom, OBS is an open source application that anybody can inspect, down to the packaging scripts. If people were concerned, they can inspect the source code and the packaging scripts themselves.

I appreciate the concern on this issue, but as a matter or practicality I don't think we will have anything to worry about vis-à-vis public perception.

Reply:
> IMO this is absolutely miserable UX.

100% agree. But it's miserable UX in the name of security, which is basically the whole story of macOS Catalina 😂

Reply:
> I appreciate the concern on this issue, but as a matter or practicality I don't think we will have anything to worry about vis-à-vis public perception.

I think transparency of what is done with elevated privileges in a .pkg can be highlighted not only implicitly by the mere fact that OBS is open source software which can be inspected by everyone, but also explicitly by clearly stating why the privileges are required or what they are used for in the installer itself.

That would help people like me who rather unpack the installer binary data themselves to extract the App and inspect the scripts rather than risking it.

Since the installer usually just contains the App, maybe it would even be feasible to allow downloading that in case people are not interested in providing root access for a given featureset, or simply are not root on their computers, certainly the case for many corporate installations of MacOS.

Reply:
Could we split things up to have an install dmg with the app and a separate small pkg for the stuff that needs to be in a pkg? People who just want the app can still just drag the app into the Applications folder. If we make it clear what is in the optional pkg, people who want it can install the pkg.

Reply:
Personal opinion: I feel that macOS is already enough of a challenge to maintain (the last few major/minor releases of OBS have had delayed macOS versions for various reasons) that introducing complexity to packaging, distribution, and even providing a secondary method of installing components required for the virtual camera will put unnecessary strain & stress on Jim, DDRBoxman, and others who are tasked with maintaining OBS on macOS.

Additionally, as dodgepong mentioned, we distributed a .pkg for years before without complaint. 

The installer will be clear about what it's doing, allowing users who just want the software to work to use it. If the user wants to avoid the installer, they are welcome to build OBS themselves.

Reply:
To elaborate a bit more on the risks: the CoreMediaIO DAL plugin that implements the virtual camera runs in-process with other apps that use the virtual camera, whether or not you're currently using the virtual camera as a source. This presents a security risk to those apps, since the plugin could theoretically exfiltrate the app's data. This is why some apps (like Zoom) no longer run DAL plugins out of the box. Some users may not feel comfortable with this risk, though I agree with @dodgepong that since OBS is open source and auditable it's maybe not a huge concern.

But running in-process also presents a stability risk: if the OBS DAL code crashes, then the host app may crash. Some apps (like Chrome) load their DAL plugins in a different process which avoids crashing all of Chrome when a plugin crashes, but it causes none of the other cameras to work (I've had to restart Chrome and rejoin many meetings lately because my janky in-progress DAL code crashed!). Other apps (e.g. QuickTime) just crash. This means that if we make a mistake in OBS, we have the potential to break our users' software and/or webcams, even when they're not using OBS at all, which would be the most miserable UX! Since it's non-obvious how to uninstall the DAL plugin (would require a script or terminal command) a mistake here could be a huge support headache as well.

That's why I don't think OBS should install a DAL plugin by default for all users. I think it should be an opt-in thing users can choose to install or gets installed automatically when the user first tries to use it. My stance would change if we determine that the majority of OBS users use the virtual camera, but I think it'd be prudent to wait and see if that's true before installing the DAL plugin by default for all OBS users.

> @jasongrout: Could we split things up to have an install dmg that contains both the app and a separate small pkg for the stuff that needs to be in a pkg?

I like this as a solution! It would probably also be the easiest to implement since it would be mostly just copying my scripts/actions verbatim from my codebase into the OBS codebase. I've definitely seen this pattern in apps before when they need to come with drivers.

> @Byron: maybe it would even be feasible to allow downloading that in case people ... are not root on their computers, certainly the case for many corporate installations of MacOS.

This is a good consideration I didn't think of! Some people may not be able to install `.pkg` files.

Reply:
Although the RFC is merged, I would like to state for the record that I personally do not like .pkg files.  They require extra installer signing which we lacked the entire time, and caused many users to just not be able to install it because they couldn't figure out that the only way to get it to work was to right-click and go to open to bypass that.  I know that we can probably get installer signing, but I personally really like the process with .dmg.  It's a much more simple install process that apple users are familiar with.  I don't know about the other issues with .pkg files stated above, but if those are an issue then that's something to seriously consider as well.

> I feel that macOS is already enough of a challenge to maintain

Personally, I think the .pkg files are just more stuff to maintain.  A .dmg file is much less.

What should we do though?  Perhaps there can be a compromise, such as having a dialog requesting installation only when the user first tries to use the virtualcam.  But ultimately, I'm unsure of what the best thing to do is here, because both have their downsides.

I just really like the .dmg setup because it's so simple and elegant, and doesn't require installation of third-party packages right off the bat.  It would be a bummer if we had to go back to .pkg files.

But conversely, if you asked me to do the same thing (ask to install on first use) on windows, I would be shocked and appalled.  I'd want to put everything in the installer.  Maybe it's just the way apple treats installers compared to windows.  Installers feel less desirable on macs though.

I don't know.  I have no idea what the right thing to do is.  Apple feels so backwards and restrictive to me.

Reply:
FWIW the right thing is to distribute a `.dmg` (especially due to codesigning and notarisation requirements) and if installation with elevated privileges is needed, handle this as part of the app's onboarding - this is common e.g. for file managers that require a helper daemon for elevated file access. Using `pkg` for apps has in essence become a red flag for users, because "official" apps are always packaged in `dmg` and don't require elevated privileges or installations, `pkg`s thus have become "sketchy".

The desired UX flow would be to copy OBS to the Applications directory, launch it, handle the onboarding permissions screen that we might have some day which includes a button to install the DAL plugin, which _then_ asks for elevated permissions (which we also told the user about on the screen).

As a user might dismiss this screen, we'd need a way to install it after anyway, which could be a menu bar item as suggested. This once again would trigger an elevation prompt and install the plugin as desired. This is also common in other apps, so not an "alien" concept.

Reply:
The issue there is that we don't really have much onboarding to speak of for Mac, other than the one-time prompt to run the auto-config, so that would have to be developed almost from scratch.

For what it's worth, we really _should_ already have an onboarding flow for Mac to do things like request permission to capture the microphone/screen/camera.

Reply:
Agreed, what I tried to get at was that the menu bar item would be needed anyway and thus would still be a necessary implementation detail.

Reply:
Ah, fair point then. Do you think an entry in the Tools menu is appropriate for this? Or do you have any other ideas of where it would best go?

Reply:
Yeah "Tools" seems to be the current home for all additional functionality and plugins. Simplest implementation might be to have the current single item do the additional checks and installation of the DAL plugin and keeps the "Enable" state if it fails or a user cancels the process.

There might be more ways to make this menu item more verbose about state of the plugin, but I have no idea if it's feasible using Qt.

Reply:
@johnboiles I've been working on code for an onboarding flow and also investigated how to install/update the DAL plugin at runtime without the need for an installer with elevated privileges.

The short story is that it's possible with a little "hack" as OBS is not a sandboxed app and thus the usual ways to get elevated write permissions for the target folders do not work (e.g. show a file browsing dialog defaulting to the target dir, which would give write permissions for a file operations).

The only way to make this work is to invoke an AppleScript that runs a `cp -R` with elevated privileges. It _does_ work without issue though.

Now the pickle: We _could_ have this code run as part of the plugin when it's being activated, then remove it later once we have a proper macOS onboarding flow. It would allow release of this plugin earlier, but introduces a small amount of maintenance later.

This in essence does the deed:

https://github.com/PatTheMav/obs-studio/blob/bd849551e6dcd8c525e60e35d1f5d046a39ca53c/UI/platform-osx.mm#L153

When OBS is run as a bundled app, the permission dialog will look "proper", alas there is no way to add an explanation _what_ OBS needs it for, but I hope that triggering this through the menu item will make it clear (excuse the German-ness of it):

![virtualcam-dal-permission](https://user-images.githubusercontent.com/9436503/90232654-49b15e00-de1d-11ea-8750-1c594a3ad11c.jpg)

The DAL plugin is copied dutifully to where it should be and I checked that e.g. Quicktime picked up the virtual cam output directly after copying over the plugin (no restarts of OBS or the system needed).